src/Monarch/VirtualDom/Patch.ts

```diff
@@ -0,0 +1,174 @@
+    export interface Redraw extends Tagged<typeof Redraw> {
+        vNode: VirtualDomTree<any>
+    }
+    /**
+     * Smart data constructor for `Redraw` type
+     */
+    export function mkRedraw(vNode: VirtualDomTree<any>): Redraw {
+        return { tag: Redraw, vNode }
+    }
+
+    // SUM TYPE: Text
+
+    /**
+     * `Text` tag
+     *
+     * Use it for pattern matching.
+     */
+    export const Text = Tag.Text
+    /**
+     * `Text` type constructor
+     */
+    export interface Text extends Tagged<typeof Text> {
+        text: string
+    }
+    /**
+     * Smart data constructor for `Text` type
+     */
+    export function mkText(text: string): Text {
+        return { tag: Text, text }
+    }
+
+    // SUM TYPE: Facts
+
+    /**
+     * `Facts` tag
+     *
+     * Use it for pattern matching.
+     */
+    export const Facts = Tag.Facts
+    /**
+     * `Facts` type constructor
+     */
+    export interface Facts extends Tagged<typeof Facts> {
+        diff: OrganizedFacts
+    }
+    /**
+     * Smart data constructor for `Facts` type
+     */
+    export function mkFacts(diff: OrganizedFacts): Facts {
+        return { tag: Facts, diff }
+    }
+
+    // SUM TYPE: RemoveFromEnd
+
+    /**
+     * `RemoveFromEnd` tag
+     *
+     * Use it for pattern matching.
+     */
+    export const RemoveFromEnd = Tag.RemoveFromEnd
+    /**
+     * `RemoveFromEnd` type constructor
+     */
+    export interface RemoveFromEnd extends Tagged<typeof RemoveFromEnd> {
+        delta: number
+    }
+    /**
+     * Smart data constructor for `RemoveFromEnd` type
+     */
+    export function mkRemoveFromEnd(delta: number): RemoveFromEnd {
+        return { tag: RemoveFromEnd, delta }
+    }
+
+    // SUM TYPE: Appeend
+
+    /**
+     * `Append` tag
+     *
+     * Use it for pattern matching.
+     */
+    export const Append = Tag.Append
+    /**
+     * `Append` type constructor
+     */
+    export interface Append extends Tagged<typeof Append> {
+        children: readonly VirtualDomTree<any>[]
+        from: number
+    }
+    /**
+     * Smart data constructor for `Append` type
+     */
+    export function mkAppend<message>(children: readonly VirtualDomTree<message>[], from = 0): Append {
+        return { tag: Append, children, from }
+    }
+}
+
+export function unsafe_applyPatch(domNode: Node, patch: Patch): void {
+    switch (patch.tag) {
+        case Patch.Text:
+            return unsafe_applyTextPatch(<Text>domNode, patch)
+
+        case Patch.Redraw:
+            return unsafe_applyRedrawPatch(domNode, patch)
+
+        case Patch.Facts:
+            return unsafe_applyFacts(domNode, patch.diff)
+
+        case Patch.RemoveFromEnd:
+            return unsafe_applyRemoveFromEndPatch(domNode, patch)
+
+        case Patch.Append:
+            return unsafe_applyAppendPatch(domNode, patch)
+    }
+}
+
+function unsafe_applyTextPatch(textNode: Text, { text }: Patch.Text): void {
+    textNode.replaceData(0, text.length, text)
```

Comment:
I think this line is wrong. Let's say we have
```js
const textNode = document.createTextNode("Hello World");
const newText = "text";
```
diff will create text patch and call this function
```js
textNode.replaceData(0, newText.length, newText);
```
and textNode will be "testo World"

Reply:
Right. Let's play with old text.